asserts/snapasserts/snapasserts.go

```diff
@@ -0,0 +1,162 @@
+		//   (more likely if it is snap-revision not matching)
+		//   people would need to report this
+		return fmt.Errorf("snap %q does not have expected ID or revision according to assertions (metadata is broken or tampered): %s / %s != %d / %s", name, si.Revision, snapID, snapRev.SnapRevision(), snapRev.SnapID())
+	}
+
+	snapDecl, err := findSnapDeclaration(snapID, name, db)
+	if err != nil {
+		return err
+	}
+
+	if snapDecl.SnapName() != name {
+		return fmt.Errorf("cannot install snap %q that is undergoing a rename to %q", name, snapDecl.SnapName())
+	}
+
+	return nil
+}
+
+// ReconstructSideInfoFlags represent the mode flags for ReconstructSideInfo.
+type ReconstructSideInfoFlags int
+
+const (
+	// AllowWithoutSignatures flags allowing extracting information from a snap for which signatures cannot be found.
+	AllowWithoutSignatures ReconstructSideInfoFlags = iota + 1
+)
+
+// ReconstructSideInfo tries to construct a SideInfo for the given snap using its digest to find the relevant snap assertions with the information in the given database. It will fail if it cannot find them, unless flags is AllowWithoutSignatures in which case it extracts unsafely just the name from the snap itself.
+func ReconstructSideInfo(snapPath string, flags ReconstructSideInfoFlags, db asserts.RODatabase) (*snap.SideInfo, error) {
+	snapSHA3_384, snapSize, err := asserts.SnapFileSHA3_384(snapPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// get relevant assertions and reconstruct metadata
+	a, err := db.Find(asserts.SnapRevisionType, map[string]string{
+		"snap-sha3-384": snapSHA3_384,
+	})
+	if err == asserts.ErrNotFound {
+		if flags == AllowWithoutSignatures {
+			// unsafe fallback mode
+			return unsafeReconstructSideInfo(snapPath)
+		}
+		return nil, fmt.Errorf("cannot find signatures with metadata for snap %q", snapPath)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	snapRev := a.(*asserts.SnapRevision)
+
+	if snapRev.SnapSize() != snapSize {
+		return nil, fmt.Errorf("snap %q does not have expected size according to signatures (broken or tampered): %d != %d", snapPath, snapSize, snapRev.SnapSize())
+	}
+
+	snapID := snapRev.SnapID()
+
+	snapDecl, err := findSnapDeclaration(snapID, snapPath, db)
+	if err != nil {
+		return nil, err
+	}
+
+	name := snapDecl.SnapName()
+
+	// TODO: once we are fully migrated to assertions this can
+	// be done dynamically later instead of statically here
+	a, err = db.Find(asserts.AccountType, map[string]string{
+		"account-id": snapRev.DeveloperID(),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("internal error: cannot find developer account for snap %q (%q): %s", name, snapPath, snapRev.DeveloperID())
+	}
+	devAcct := a.(*asserts.Account)
+
+	return &snap.SideInfo{
+		RealName:    name,
+		SnapID:      snapID,
+		Revision:    snap.R(snapRev.SnapRevision()),
+		DeveloperID: snapRev.DeveloperID(),
+		Developer:   devAcct.Username(),
+	}, nil
+}
+
+func unsafeReconstructSideInfo(snapPath string) (*snap.SideInfo, error) {
```

Comment:
we are already opening the snap multiple times atm (once in daemon and a couple in snapstate),  in daemon essentially to just have the name early to use in messages. The idea was to have one consistent function to use where we have readSnapInfo now in daemon/api.go atm